### PR TITLE
Improve event generator performance

### PR DESCRIPTION
**Pull request by Aaron1011**
SpongeAPI | [SpongeCommon](https://github.com/SpongePowered/SpongeCommon/pull/1201) | [event-impl-gen](https://github.com/SpongePowered/event-impl-gen/pull/2)

Tl;dr: Sponge currently generates the same constructor for all dynamically created event implementations - one that takes a `Map` of property names to values However, constantly creating and accessing maps can be a drain on performance when many events are being fired. This PR modifiers the event generator to create proper constructors for every event class.

**Background**:

Event generation in Sponge currently consists of two parts. The user-facing `SpongeEventFactory` class is generated by [event-impl-gen](https://github.com/SpongePowered/event-impl-gen), which generates the proper static methods from the existing event interfaces. `SpongeEventFactory` then uses [ClassGenerator](https://github.com/SpongePowered/SpongeAPI/blob/bleeding/src/main/java/org/spongepowered/api/util/generator/event/factory/ClassGenerator.java) to generate a class implementing the desired event interface. Parameters are passed from the static method in `SpongeEventFactory` to the generated event constructor by storing them in a `Map`. The generated class reads the necessary parameters from this `Map` to populate its fields.

**The issue**:

While this system is simple, it suffers from a lot of overhead due to the need to construct, populate, and read from a `Map` every time an event is created. On large servers, where events are constantly being fired, this can lead to performance issues.

To solve this, I've modified both parts of the event generation to avoid the need for an intermediate `Map`.

On the `event-impl-gen` side, I now generate an interface called `SpongeEVentFactoryImpl` (final name TBD). `SpongeEventFactory` now delegates all methods to this interface, by calling `Sponge.getEventFactory`. This prevents the need to use any bytecode modification directly on `SpongeEventFactory` itself, as we can instead dynamically generate a class to implement `SpongeEventFactoryImpl`.

On the `ClassGenerator` side, I dynamically generate a class to implement `SpongeEventFactoryImpl`. Each method in this class invokes the actual constructor of the generated event implementation class, with no `Map` indirection neded.

**Caveats**:

Generating the implementation for `SpongeEventFactoryImpl` (done in `FactoryInterfaceGenerator`) relies on the method parameter names being available. Some of the information needed to sort parameters of methods in `SpongeEventFactory/SpongeEventFactoryImpl` into the proper order (`Cause` always comes first, `originalXXX` and `xxx` go together) comes from SpongeAPI's `build.gradle`. In order to avoid copying this information into other parts of SpongeAPI, I use the method parameter names to convert the user-friendly parameter ordering into a simple alphabetical ordering (for use in `ClassGenerator`).

Accessing method parameter names through reflection requires the user of the `-parameters` javac flag. While it's simple to specify this in Gradle, IDEs might not always use this flag. If you receive an error such as `Parameter <param> for method <method> has no name!`, try enabling this flag in your IDE compiler settings.

### CODE REVIEW

**Comment by DDoS** [comment]
`SpongeEventFactory` seems redundant now. I guess it's easier to keep for backwards compatibility, but calling the `SpongeEventFactoryImpl` methods directly should be encouraged. It should be `@Deprecated`. BTW, `Impl` for an interface is a terrible name (I know it's temporary).

**Comment by gabizou** [comment]
In my honest opinion, we should have the new `SpongeEventFactoryImpl` renamed to something more permanent since a change between now and 6.0 is going to be a binary breakage anyways (as changing the name back to `SpongeEventFactory` will result in binary breakages). I'd much rather it be called `EventFactory` now (since that class was unused before) and not have to change the class name in the future (plugins using their own factory already have their own factory names anyways).

**Comment by stephan-gh** [comment]
I still think that generating the bytecode directly at compile time using Gradle is a better idea than requiring additional compiler arguments in the IDE. Re-configuring the IDE takes at least as much time as executing a simple task once to generate the event factory.

The advantage of generating bytecode directly at compile time is that we no longer get merge conflicts in the generated source code. We could also keep using the static methods of the event factory, because the generated factory class could directly contain the implementations without the need for an additional interface.

File: src/main/java/org/spongepowered/api/util/generator/event/factory/ClassGenerator.java [hunk 1]
```diff
@@ -318,31 +342,10 @@ private void generateConstructor(ClassWriter classWriter, String internalName, C
...
-            mv.visitMethodInsn(INVOKEVIRTUAL, "java/lang/StringBuilder", "append", "(Ljava/lang/String;)Ljava/lang/StringBuilder;", false);
-            mv.visitVarInsn(ALOAD, 1);
-            mv.visitMethodInsn(INVOKEINTERFACE, "java/util/Map", "keySet", "()Ljava/util/Set;", true);
-            mv.visitMethodInsn(INVOKEVIRTUAL, "java/lang/StringBuilder", "append", "(Ljava/lang/Object;)Ljava/lang/StringBuilder;", false);
-            mv.visitMethodInsn(INVOKEVIRTUAL, "java/lang/StringBuilder", "toString", "()Ljava/lang/String;", false);
-            mv.visitMethodInsn(INVOKESPECIAL, "java/lang/IllegalArgumentException", "<init>", "(Ljava/lang/String;)V", false);
-            mv.visitInsn(ATHROW);
-
-            mv.visitLabel(afterException);
+            if (type.equals(Type.LONG_TYPE) || type.equals(Type.DOUBLE_TYPE)) {
```
Comment (Mumfrey) [inline src/main/java/org/spongepowered/api/util/generator/event/factory/ClassGenerator.java]: `if (type.getSize() > 1)` would be better here.

**Comment by bloodmc** [comment]
@Aaron1011 What is status of this PR?

**Comment by Aaron1011** [comment]
@Minecrell: While we could generate the bytecode at compile time, I feel like it's nicer to have the project be immediately compileable (without the need to run any gradle tasks). It's also nice to be able to link directly to the event factory on Github.

I think I may merge this as-is - but we can always decide to change how we generate the implememention classes later, since it wouldn't require an API break.

**Pull request closed without merging**